### PR TITLE
Add registry based version info to telemetry

### PR DESCRIPTION
**Pull request by DaveTryon**
#### Describe the change
Our telemetry pipeline (Application Insights) reports OS version data fetched via an API that intentionally lies about windows versions unless certain manifest flags are set. Unfortunately, those same flags also alter high DPI behavior, which would be a high impact change. Workaround is to add our own windows version information to telemetry, using Registry keys that work at least as far as back as Win7, and that are likely to be around for a very long time. This change may complicate the work to split the Win32 assembly into 2 pieces, but since that work hasn't yet started, there should be no conflict in making this change now.

#### PR checklist

- [ ] Run through of all [test scenarios](https://github.com/Microsoft/accessibility-insights-windows/blob/master/docs/Scenarios.md) completed?
- [x] Does this address an existing issue? If yes, Issue# - #326
- [ ] Includes UI changes?
  - [ ] Run the production version of Accessibility Insights for Windows against a version with changes.
  - [ ] Attach any screenshots / GIF's that are applicable.

> Note: After the PR has been created, certain checks will be kicked off. All of these checks must pass before a merge.

### CODE REVIEW

File: src/AccessibilityInsights/AccessibilityInsights.csproj [hunk 1]
```diff
@@ -262,6 +262,10 @@
       <Project>{eee9cd52-ca10-41ae-ac6e-98ae4bbdfa2d}</Project>
       <Name>CommonUxComponents</Name>
     </ProjectReference>
+    <ProjectReference Include="..\AccessibilityInsights.Extensions.Telemetry\Extensions.Telemetry.csproj">
```
Comment (DaveTryon) [inline src/AccessibilityInsights/AccessibilityInsights.csproj]: Durp. This was supposed to stay local and not get pushed

File: src/AccessibilityInsights/MainWindow.xaml.cs [hunk 2]
```diff
@@ -197,7 +197,13 @@ private static void InitTelemetry()
             Logger.AddOrUpdateContextProperty(TelemetryProperty.Version, VersionTools.GetAppVersion());
             Logger.AddOrUpdateContextProperty(TelemetryProperty.AppSessionID, Guid.NewGuid().ToString());
             Logger.AddOrUpdateContextProperty(TelemetryProperty.SessionType, "Desktop");
-            Logger.AddOrUpdateContextProperty(TelemetryProperty.ReleaseChannel, ConfigurationManager.GetDefaultInstance().AppConfig.ReleaseChannel.ToString());
+            Logger.AddOrUpdateContextProperty(TelemetryProperty.WindowsVersion, ConfigurationManager.GetDefaultInstance().AppConfig.ReleaseChannel.ToString());
```
Comment (jalkire) [inline src/AccessibilityInsights/MainWindow.xaml.cs]: Looks like TelemetryProperty.WindowsVersion and TelemetryProperty.ReleaseChannel got flipped between here and line 205
```suggestion
            Logger.AddOrUpdateContextProperty(TelemetryProperty.ReleaseChannel, ConfigurationManager.GetDefaultInstance().AppConfig.ReleaseChannel.ToString());
```

File: src/AccessibilityInsights/MainWindow.xaml.cs [hunk 3]
```diff
@@ -197,7 +197,13 @@ private static void InitTelemetry()
...
+
+            string windowsVersion = NativeMethods.GetCurrentWindowsVersionForTelemetry();
+            if (!string.IsNullOrEmpty(windowsVersion))
+            {
+                Logger.AddOrUpdateContextProperty(TelemetryProperty.ReleaseChannel, windowsVersion);
```
Comment (jalkire) [inline src/AccessibilityInsights/MainWindow.xaml.cs]:
```suggestion
                Logger.AddOrUpdateContextProperty(TelemetryProperty.WindowsVersion, windowsVersion);
```

Comment (DaveTryon) [reply src/AccessibilityInsights/MainWindow.xaml.cs]: Oops! I changed the wrong one

File: src/AccessibilityInsights/MainWindow.xaml.cs [hunk 4]
```diff
@@ -198,6 +198,12 @@ private static void InitTelemetry()
             Logger.AddOrUpdateContextProperty(TelemetryProperty.AppSessionID, Guid.NewGuid().ToString());
             Logger.AddOrUpdateContextProperty(TelemetryProperty.SessionType, "Desktop");
             Logger.AddOrUpdateContextProperty(TelemetryProperty.ReleaseChannel, ConfigurationManager.GetDefaultInstance().AppConfig.ReleaseChannel.ToString());
+
+            string windowsVersion = NativeMethods.GetCurrentWindowsVersionForTelemetry();
+            if (!string.IsNullOrEmpty(windowsVersion))
```
Comment (RobGallo) [inline src/AccessibilityInsights/MainWindow.xaml.cs]: Should we log something if we can't get the version?

File: src/AccessibilityInsights.Win32Tests/Win32ApisTests.cs [hunk 5]
```diff
@@ -263,5 +263,47 @@ public void IsWindowsRS5OrLater_BuildValueIsGreaterThanTarget_ReturnsTrue()
...
+        [TestMethod]
+        public void GetCurrentWindowsVersionForTelemetry_CurrentVersionDoesNotExist_ReturnsNull()
+        {
+            using (ShimsContext.Create())
+            {
+                List<string> returnValues = new List<string> { "" };
+                int index = 0;
+                ShimRegistry.GetValueStringStringObject = (_, __, def) => (++index > returnValues.Count ? def : returnValues[index - 1]);
+
+                Assert.IsNull(NativeMethods.GetCurrentWindowsVersionForTelemetry());
```
Comment (RobGallo) [inline src/AccessibilityInsights.Win32Tests/Win32ApisTests.cs]: Not that I expect any change in this PR, but isn't the NativeMethods class supposed to only be for pInvoke signatures of native methods?

**Comment by DaveTryon** [comment]
@neocln : The AppInsights code already uses Environment,OSVersion internally. This is the API that misreports the OS unless the manifest file is modified. That said, I like the idea of using the AppInsights-provided flag. Having both old and new formats in the same field will generate some noise in the short run, but it will drop off over time.

**Comment by DaveTryon** [comment]
@neocln: Your proposal will cause some bumps in the short-term telemetry, as we will have multiple forms of data in the same string. That said, we can probably handle that change. I don't want to reference Win32 from the extension, so I'll just add the registry code directly into the extension. That also removes the need for touching Win32.

File: src/AccessibilityInsights.Extensions.Telemetry/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -25,7 +25,7 @@
 [assembly: Guid("d17e1749-28fe-44b9-a9f6-af8985124fe4")]
 
 #if ENABLE_SIGNING
-[assembly: InternalsVisibleTo("Axe.Windows.DesktopTests,PublicKey=002400000480000094000000060200000024000052534131000400000100010007d1fa57c4aed9f0a32e84aa0faefd0de9e8fd6aec8f87fb03766c834c99921eb23be79ad9d5dcc1dd9ad236132102900b723cf980957fc4e177108fc607774f29e8320e92ea05ece4e821c0a5efe8f1645c4c0c93c1ab99285d622caa652c1dfad63d745d6f2de5f17e5eaf0fc4963d261c8a12436518206dc093344d5ad293")]
```
Comment (DaveTryon) [inline src/AccessibilityInsights.Extensions.Telemetry/Properties/AssemblyInfo.cs]: DesktopTests doesn't actually use these internals. This was probably a remnant of a pre-refactor dependency

**Comment by DaveTryon** [comment]
Here's an event uploaded with the change--note that the client_OS field now says **6.3.18362** (the correct value for 19H1) instead of **Microsoft Windows NT 6.2.9200.0**:

image

File: src/AccessibilityInsights/MainWindow.xaml.cs (hunk 4 above)
Comment (DaveTryon) [reply src/AccessibilityInsights/MainWindow.xaml.cs]: We'll know if a problem occurs, so I don't think it's necessary

---
In reply to: [277715682](https://github.com/Microsoft/accessibility-insights-windows/pull/332#discussion_r277715682) [](ancestors = 277715682)

File: src/AccessibilityInsights.Win32Tests/Win32ApisTests.cs (hunk 5 above)
Comment (DaveTryon) [reply src/AccessibilityInsights.Win32Tests/Win32ApisTests.cs]: No longer an issue in the revised format

---
In reply to: [277719578](https://github.com/Microsoft/accessibility-insights-windows/pull/332#discussion_r277719578) [](ancestors = 277719578)

File: src/AccessibilityInsights.Win32/Win32Helper.cs [hunk 7]
```diff
@@ -193,6 +193,28 @@ internal static bool IsWindowsRS5OrLater()
             return IsAtLeastWin10WithSpecificBuild(17713); // Build 17713 is confirmed in the RS5 range
         }
 
+        /// <summary>
+        /// Get the current Windows version (intended to be used for telemetry). If trying to
```
[the remark this answers is not in the document]
Comment (DaveTryon) [reply src/AccessibilityInsights.Win32/Win32Helper.cs]: Telemetry method is now in telemetry extension

File: src/AccessibilityInsights.Extensions.Telemetry/OSHelpers.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+using Microsoft.Win32;
+
+namespace AccessibilityInsights.Extensions.Telemetry
+{
+    /// <summary>
+    /// Helpers to extract data from the OS
+    /// </summary>
+    internal static class OSHelpers
+    {
+        /// <summary>
```
Comment (RobGallo) [inline src/AccessibilityInsights.Extensions.Telemetry/OSHelpers.cs]: I still wish there were a remark as to why this is suitable for telemetry but not for features. There's no "why".

File: src/AccessibilityInsights.Extensions.TelemetryTests/OSHelpersUnitTests.cs [hunk 9]
```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```
Comment (RobGallo) [inline src/AccessibilityInsights.Extensions.TelemetryTests/OSHelpersUnitTests.cs]: Are we calling test files *UnitTests.cs" or "*Tests.cs" now?

File: src/AccessibilityInsights.Extensions.TelemetryTests/Extensions.TelemetryTests.csproj [hunk 10]
```diff
@@ -64,6 +68,7 @@
   <ItemGroup>
     <Compile Include="AITelemetryTests.cs" Condition="$(FAKES_SUPPORTED) == 1" />
     <Compile Include="Properties\AssemblyInfo.cs" />
+    <Compile Include="OSHelpersUnitTests.cs" />
```
Comment (RobGallo) [inline src/AccessibilityInsights.Extensions.TelemetryTests/Extensions.TelemetryTests.csproj]: I don't see us excluding the fakes based on the environment variable anywhere. We can do this here at the file level, or at the function level in the file itself.

Comment (DaveTryon) [reply src/AccessibilityInsights.Extensions.TelemetryTests/Extensions.TelemetryTests.csproj]: This was added in the most recent iteration

File: src/AccessibilityInsights.Extensions.TelemetryTests/OSHelpersUnitTests.cs (hunk 9 above)
Comment (DaveTryon) [reply src/AccessibilityInsights.Extensions.TelemetryTests/OSHelpersUnitTests.cs]: The Unit Tests documentation says *UnitTests.cs. The code to run tests looks for *Tests.dll

File: src/AccessibilityInsights.Extensions.Telemetry/OSHelpers.cs (hunk 8 above)
Comment (DaveTryon) [reply src/AccessibilityInsights.Extensions.Telemetry/OSHelpers.cs]: There are countless blog pages on the problems with the various forms of GetVersion. I don't see a lot of value in adding yet another commentary.

**Pull request merged**